### PR TITLE
client: LAZY_IO support

### PR DESCRIPTION
**Pull request by ukernel**
'client_force_lazyio' option enables LAZY_IO gloablly for libcephfs
and ceph-fuse mount.

'ceph_lazyio(...)' and 'ceph_ll_lazyio(...)' enable LAZY_IO for open
file in libcephfs.

'ioctl(fd, CEPH_IOC_LAZYIO, 1UL)' enables LAZY_IO for open fils in
ceph-fuse mount.

Fixes: http://tracker.ceph.com/issues/20598
Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

### CODE REVIEW

**Comment by ukernel** [comment]
@dvanders   are the ways to enable lazyio OK?

**Comment by cernceph** [comment]
LGTM. Thanks!!!

**Comment by dvanders** [comment]
LGTM. Thanks !!!

File: src/include/ceph_fs.h [hunk 1]
```diff
@@ -396,6 +396,7 @@ extern const char *ceph_mds_op_name(int op);
 #define CEPH_O_CREAT           00000100
 #define CEPH_O_EXCL            00000200
 #define CEPH_O_TRUNC           00001000
+#define CEPH_O_LAZY            00020000
```
Comment (batrick) [inline src/include/ceph_fs.h]: I'm okay with a function to enable LAZYIO (e.g. `ceph_lazyio`) and a blanket config but I think we should forgo this open flag `CEPH_O_LAZY`. It's not standard and because it was unusable (can't modify existing apps), LAZYIO never got traction years ago.

**Comment by batrick** [comment]
@ukernel PTAL

File: src/client/Client.cc [hunk 2]
```diff
@@ -8475,7 +8510,6 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
 	ldout(cct, 8) << "Unable to get caps after open of inode " << *in <<
 			  " . Denying open: " <<
 			  cpp_strerror(result) << dendl;
-	in->put_open_ref(cmode);
```
[the remark this answers is not in the document]
Comment (ukernel) [reply src/client/Client.cc]: it's bug

**Comment by ukernel** [comment]
On Sat, Jun 23, 2018 at 7:19 AM Patrick Donnelly
<notifications@github.com> wrote:
>
> @batrick requested changes on this pull request.
>
> Few comments.
>
> Also, what happened to using xattrs on a subtree to enable LAZYIO?
>
> ________________________________
>
>

semantic of subtree lazy io is not clear:

1. Who checks the xattrs, client or mds? If client does the check, it
needs to check ancestor inodes' xattrs (this is expensive). If mds
does the check, should it issue Fl to clients blindly?

2. If the xattr is set, should lazy io be enabled automatically or
still require explicit instruction from user?

**Comment by batrick** [comment]
> If mds does the check, should it issue Fl to clients blindly?

Yes. The user has already made the consistency requirements clear via the xattr.

>  2. If the xattr is set, should lazy io be enabled automatically or still require explicit instruction from user?

Automatically.

File: src/client/Client.cc [hunk 3]
```diff
@@ -9232,7 +9273,8 @@ int64_t Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
     }
   }
 
-  if (cct->_conf->client_oc && (have & CEPH_CAP_FILE_BUFFER)) {
+  if (cct->_conf->client_oc &&
+      (have & (CEPH_CAP_FILE_BUFFER | CEPH_CAP_FILE_LAZYIO))) {
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/client/Client.cc]: I don't like that `Fl` means that all writes/reads become synchronous. That is unnecessary and will hurt performance (which this feature is all about!). The need for two clients doing lazy I/O to have cache coherency will be the exception, not the norm. Let's change it so that the MDS can hand out `Fcbl` in this PR and, if it's easy, add another subtree xattr to have the MDS to **not** give out `Fcb` with `Fl`, if the user needs that.

File: src/client/Client.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (batrick) [reply src/client/Client.cc]: Separate commit/PR/tracker then, please.

**Comment by ukernel** [comment]
On Wed, Jun 27, 2018 at 9:15 PM Patrick Donnelly <notifications@github.com>
wrote:

> If mds does the check, should it issue Fl to clients blindly?
>
> Yes. The user has already made the consistency requirements clear via the
> xattr.
>

>    1. If the xattr is set, should lazy io be enabled automatically or
>    still require explicit instruction from user?
>
> Automatically.
>
The problem is that some user may not know the xattr is set, they will be
confusing when they found inconsistance.

File: src/client/Client.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (ukernel) [reply src/client/Client.cc]: what's benefits of doing this. In single client case, mds issues Fcb to client even it wants Fl.

[the remark this answers is not in the document]
Comment (batrick) [reply src/client/Client.cc]: > what's benefits of doing this.

So I wasn't thinking right. Forget what I said "I don't like that Fl means that all writes/reads become synchronous".

> In single client case, mds issues Fcb to client even it wants Fl.

I'm wondering if this could be changed so that `Fl` simply means that other clients may be doing reads/writes with no cache coherency guarantees. The MDS automatically hands out `Fb` with the `Fwl` cap... because the client **can** buffer writes.

In other words, for the client is there a real difference in the code between `Fwb` and `Fwl`. What if the MDS always gives you `Fb` with `Fl`? I feel this could be simpler and the `Fl` cap could just be checked by some `ceph_is_cache_coherent(int fd)` function.

/late night thinking. Please let me know if that doesn't make sense.

[the remark this answers is not in the document]
Comment (ukernel) [reply src/client/Client.cc]: Fb and Fl have difference.  Fb means client can do buffered write, Fl means client can do buffered write for files opened in lazy IO mode.

Issue Fb with Fl may make client code a little simpler. But it will mds code more complex. (move adjust_caps_used_for_lazyio logical to mds)

[the remark this answers is not in the document]
Comment (batrick) [reply src/client/Client.cc]: > Fb and Fl have difference. Fb means client can do buffered write, Fl means client can do buffered write for files opened in lazy IO mode.

Logically only the MDS cares about LAZYIO as it permits handing out `Fbc` to more than one client in the presence of multiple writers. The client does not, itself, need to use LAZYIO cap at all except to ask the MDS to allow LAZYIO. However, keeping LAZYIO as a cap still makes sense to simplify code in the MDS; however, it does not modify the client's behavior.

> Issue Fb with Fl may make client code a little simpler. But it will make mds code more complex. (move adjust_caps_used_for_lazyio logical to mds)

Is it really that complex though? We're just going to (a) add `Fb` if `Flw` and (b) add `Fc` if `Flr`.

We do not need to worry about backwards-compatibility so I would prefer we do this in a way that makes sense.

**Comment by ukernel** [comment]
On Fri, Jun 29, 2018 at 4:57 AM Patrick Donnelly <notifications@github.com>
wrote:

> *@batrick* commented on this pull request.
> ------------------------------
>
> In src/client/Client.cc
> <https://github.com/ceph/ceph/pull/22450#discussion_r198983431>:
>
> > @@ -9232,7 +9273,8 @@ int64_t Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
>      }
>    }
>
> -  if (cct->_conf->client_oc && (have & CEPH_CAP_FILE_BUFFER)) {
> +  if (cct->_conf->client_oc &&
> +      (have & (CEPH_CAP_FILE_BUFFER | CEPH_CAP_FILE_LAZYIO))) {
>
> Fb and Fl have difference. Fb means client can do buffered write, Fl means
> client can do buffered write for files opened in lazy IO mode.
>
> Logically only the MDS cares about LAZYIO as it permits handing out Fbc
> to more than one client in the presence of multiple writers. The client
> does not, itself, need to use LAZYIO cap at all except to ask the MDS to
> allow LAZYIO. However, keeping LAZYIO as a cap still makes sense to
> simplify code in the MDS; however, it does not modify the client's behavior.
>
For example:

A client first does lazy IO, then it does normal IO.  when it does normal
IO, it may still have Fl (wanted caps updated lazily) .  We can check 'Fb
and not Fl'



> Issue Fb with Fl may make client code a little simpler. But it will make
> mds code more complex. (move adjust_caps_used_for_lazyio logical to mds)
>
> Is it really that complex though? We're just going to (a) add Fb if Flw
> and (b) add Fc if Flr.
>
>
The state table in src/mds/locks.c can't support this directly, need more
code (at the places that issues caps and handles cap release ) to deal with
this special case.


We do not need to worry about backwards-compatibility so I would prefer we

File: src/client/Client.cc [hunk 4]
```diff
@@ -8425,9 +8458,11 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
   }
 
   // use normalized flags to generate cmode
-  int cmode = ceph_flags_to_mode(ceph_flags_sys2wire(flags));
-  if (cmode < 0)
-    return -EINVAL;
+  int cflags = ceph_flags_sys2wire(flags);
+  if (cct->_conf->get_val<bool>("client_force_lazyio"))
```
Comment (batrick) [inline src/client/Client.cc]: @ukernel needs rebased, needs to do `g_conf().get_val`

**Comment by ukernel** [comment]
@batrick updated

**Comment by mykaul** [comment]
[LAZY_IO Support](https://trello.com/c/Ob5au8OI/98-lazyio-support)
